store/src/main/java/org/apache/rocketmq/store/lock/AdaptiveBackOffSpinLockImpl.java

```diff
@@ -0,0 +1,193 @@
+
+    private Map<String, AdaptiveBackOffSpinLock> locks;
+
+    private final List<AtomicInteger> tpsTable;
+
+    private final List<Map<Thread, Byte>> threadTable;
+
+    private int swapCriticalPoint;
+
+    private AtomicInteger currentThreadNum = new AtomicInteger(0);
+
+    private AtomicBoolean isOpen = new AtomicBoolean(true);
+
+    public AdaptiveBackOffSpinLockImpl() {
+        this.locks = new HashMap<>();
+        this.locks.put("Reentrant", new BackOffReentrantLock());
+        this.locks.put("Spin", new BackOffSpinLock());
+
+        this.threadTable = new ArrayList<>(2);
+        this.threadTable.add(new ConcurrentHashMap<>());
+        this.threadTable.add(new ConcurrentHashMap<>());
+
+        this.tpsTable = new ArrayList<>(2);
+        this.tpsTable.add(new AtomicInteger(0));
+        this.tpsTable.add(new AtomicInteger(0));
+
+        adaptiveLock = this.locks.get("Spin");
+    }
+
+    @Override
+    public void lock() {
+        int slot = LocalTime.now().getSecond() % 2;
+        this.threadTable.get(slot).putIfAbsent(Thread.currentThread(), Byte.MAX_VALUE);
+        this.tpsTable.get(slot).getAndIncrement();
+        boolean state;
+        do {
+            state = this.state.get();
+        } while (!state);
+
+        currentThreadNum.incrementAndGet();
+        this.adaptiveLock.lock();
+    }
+
+    @Override
+    public void unlock() {
+        this.adaptiveLock.unlock();
+        currentThreadNum.decrementAndGet();
+        if (isOpen.get()) {
+            swap();
+        }
+    }
+
+    @Override
+    public void update(MessageStoreConfig messageStoreConfig) {
+        this.adaptiveLock.update(messageStoreConfig);
+    }
+
+    @Override
+    public void swap() {
+        if (!this.state.get()) {
+            return;
+        }
+        boolean needSwap = false;
+        int slot = 1 - LocalTime.now().getSecond() % 2;
+        int tps = this.tpsTable.get(slot).get() + 1;
+        int threadNum = this.threadTable.get(slot).size();
+        this.tpsTable.get(slot).set(-1);
+        this.threadTable.get(slot).clear();
+        if (tps == 0) {
+            return;
+        }
+
+        if (this.adaptiveLock instanceof BackOffSpinLock) {
+            BackOffSpinLock lock = (BackOffSpinLock) this.adaptiveLock;
+            if (lock.getNumberOfRetreat(slot) * BASE_SWAP_LOCK_RATIO >= tps) {
+                if (lock.isAdapt()) {
+                    lock.adapt(true);
+                } else {
+                    this.swapCriticalPoint = tps * threadNum;
+                    needSwap = true;
+                }
+            } else if (lock.getNumberOfRetreat(slot) * BASE_SWAP_LOCK_RATIO * SPIN_LOCK_ADAPTIVE_RATIO <= tps) {
+                lock.adapt(false);
+            }
+            lock.setNumberOfRetreat(slot, 0);
+        } else {
+            if (tps * threadNum <= this.swapCriticalPoint * SWAP_SPIN_LOCK_RATIO) {
+                needSwap = true;
+            }
+        }
+
+        if (needSwap) {
```

Comment:
Add some comments to explain the principles and the rationale behind these numbers.

Reply:
done~